src/types.rs

```diff
@@ -80,6 +81,271 @@ impl BERDecodable for EcdsaX962<Sha256> {
+        match *self {
+            GeneralName::IpAddress(..) => true,
+            _ => false,
+        }
+    }
+
+    pub fn as_ip_address(&self) -> Option<&Vec<u8>> {
+        match self {
+            GeneralName::IpAddress(ip) => Some(ip),
+            _ => None,
+        }
+    }
+
+    pub fn into_ip_address(self) -> Option<Vec<u8>> {
+        match self {
+            GeneralName::IpAddress(ip) => Some(ip),
+            _ => None,
+        }
+    }
+
+    pub fn is_registered_id(&self) -> bool {
+        match *self {
+            GeneralName::RegisteredID(..) => true,
+            _ => false,
+        }
+    }
+
+    pub fn as_registered_id(&self) -> Option<&ObjectIdentifier> {
+        match self {
+            GeneralName::RegisteredID(oid) => Some(oid),
+            _ => None,
+        }
+    }
+
+    pub fn into_registered_id(self) -> Option<ObjectIdentifier> {
+        match self {
+            GeneralName::RegisteredID(oid) => Some(oid),
+            _ => None,
+        }
+    }
+}
+
+impl<'a> DerWrite for GeneralName<'a> {
+    fn write(&self, writer: DERWriter) {
+        match self {
+            GeneralName::OtherName(oid, tdv) =>
+                writer.write_tagged_implicit(
+                    Tag::context(0),
+                    |w| w.write_sequence(|w| {
+                        oid.write(w.next());
+                        w.next().write_tagged(Tag::context(0), |w| tdv.write(w));
+                    })),
+            GeneralName::Rfc822Name(s) =>
+                writer.write_tagged_implicit(Tag::context(1), |w| w.write_ia5_string_safe(&s)),
+            GeneralName::DnsName(s) =>
+                writer.write_tagged_implicit(Tag::context(2), |w| w.write_ia5_string_safe(&s)),
+            GeneralName::DirectoryName(n) =>
+                // explicit tagging because Name is an untagged CHOICE (X.680-0207 clause 30.6.c)
+                writer.write_tagged(Tag::context(4), |w| n.write(w)),
+            GeneralName::UniformResourceIdentifier(s) =>
+                writer.write_tagged_implicit(Tag::context(6), |w| w.write_ia5_string_safe(&s)),
+            GeneralName::IpAddress(a) =>
+                writer.write_tagged_implicit(Tag::context(7), |w| a.write(w)),
+            GeneralName::RegisteredID(oid) =>
+                writer.write_tagged_implicit(Tag::context(8), |w| oid.write(w)),
+        }
+    }
+}
+
+impl<'a> BERDecodable for GeneralName<'a> {
+    fn decode_ber(reader: BERReader) -> ASN1Result<Self> {
+        let tag_number = reader.lookahead_tag()?.tag_number;
+        if tag_number == 4 {
+            // explicit tagging because Name is an untagged CHOICE (X.680-0207 clause 30.6.c)
+            reader.read_tagged(Tag::context(4), |r| {
+                Ok(GeneralName::DirectoryName(Name::decode_ber(r)?))
+            })
+        } else {
+            reader.read_tagged_implicit(Tag::context(tag_number), |r| {
+                match tag_number {
+                    0 => {
```

Comment:
Can we have some constants for this numbers and a comment of where they come from?

Reply:
Done